toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
-#include "toolchain/driver/flags.def"
-  }
-  llvm_unreachable("All subcommands handled!");
-}
+  std::string host = llvm::sys::getDefaultTargetTriple();
+  llvm::StringRef target;
 
-auto Driver::RunHelpSubcommand(DiagnosticConsumer& /*consumer*/,
-                               llvm::ArrayRef<llvm::StringRef> args) -> bool {
-  // TODO: We should support getting detailed help on a subcommand by looking
-  // for it as a positional parameter here.
-  if (!args.empty()) {
-    ReportExtraArgs("help", args);
-    return false;
-  }
+  llvm::StringRef output_file_name;
+  llvm::StringRef input_file_name;
 
-  output_stream_ << "List of subcommands:\n\n";
+  bool asm_output = false;
+  bool force_obj_output = false;
+  bool dump_tokens = false;
+  bool dump_parse_tree = false;
+  bool dump_semantics_ir = false;
+  bool dump_llvm_ir = false;
+  bool dump_asm = false;
+  bool stream_errors = false;
+  bool preorder_parse_tree = false;
+  bool builtin_semantics_ir = false;
+};
 
-  constexpr llvm::StringLiteral SubcommandsAndHelp[][2] = {
-#define CARBON_SUBCOMMAND(Name, Spelling, HelpText) {Spelling, HelpText},
-#include "toolchain/driver/flags.def"
+struct Driver::Options {
+  static constexpr Args::CommandInfo Info = {
+      .name = "carbon",
+      // TODO: Setup more detailed version information and use that here.
+      .version = R"""(
+Carbon Language toolchain -- version 0.0.0
+)""",
+      .help = R"""(
+This is the unified Carbon Language toolchain driver. It's subcommands provide
+all of the core behavior of the toolchain, including compilation, linking, and
+developer tools. Each of these has its own subcommand, and you can pass a
+specific subcommand to the `help` subcommand to get details about is usage.
+)""",
+      .help_epilog = R"""(
+For questions, issues, or bug reports, please use our GitHub project:
+
+  https://github.com/carbon-language/carbon-lang
+)""",
   };
 
-  int max_subcommand_width = 0;
-  for (const auto* subcommand_and_help : SubcommandsAndHelp) {
-    max_subcommand_width = std::max(
-        max_subcommand_width, static_cast<int>(subcommand_and_help[0].size()));
-  }
+  enum class Subcommand {
+    Compile,
+  };
 
-  for (const auto* subcommand_and_help : SubcommandsAndHelp) {
-    llvm::StringRef subcommand_text = subcommand_and_help[0];
-    // TODO: We should wrap this to the number of columns left after the
-    // subcommand on the terminal, and using a hanging indent.
-    llvm::StringRef help_text = subcommand_and_help[1];
-    output_stream_ << "  "
-                   << llvm::left_justify(subcommand_text, max_subcommand_width)
-                   << " - " << help_text << "\n";
+  void Build(Args::CommandBuilder& b) {
+    b.AddFlag(
+        {
+            .name = "verbose",
+            .short_name = "v",
+            .help = "Enable verbose logging to the stderr stream.",
+        },
+        [&](auto& arg_b) { arg_b.Set(&verbose); });
+
+    b.AddSubcommand(CompileOptions::Info, [&](auto& sub_b) {
```

Comment:
I feel odd about all your `auto` in lambdas. Why is `sub_bo.Do` a callback instead of directly set when this lambda is called? Might be clearer with an actual type.

Reply:
Done for the lines mentioned.

Happy to switch others, although I worry about the loss of brevity for most of the argument cases. That said, maybe it makes sense to try some of the argument API improvements first and then see if that or a better name or some other change make the most sense.